### ISSUE TITLE
mix clean --deps does not clean out .erl files generated by yecc/leex

### ISSUE THREAD

**Issue by f355** [elixir-lang/elixir#8262]
### Current behavior

When running `mix clean --deps`, the `.erl` files generated from `.xrl`/`.yrl` are left behind.

### Expected behavior

All artefacts and intermediate files for the dependencies are deleted after running `mix clean --deps`, leaving the tree as it looked after `mix deps.get`, before the compilation.

### Steps to reproduce

I'll use Absinthe as an example dependency. I'm using a GitHub tag instead of Hex version, because the artefact pushed to Hex has the `.erl` files in question included for some reason (which might be a separate issue around `mix hex.publish` that's outside of the scope for this bug report).

- Create a new project: `mix new mix_bug && cd mix_bug`
- Edit mix.exs to add `{:absinthe, git: "https://github.com/absinthe-graphql/absinthe.git", tag: "v1.4.13"}` to the list of deps.
- Fetch the dependencies: `mix deps.get`
- Verify there are no `.erl` files:
```bash
$ ls deps/absinthe/src
absinthe_lexer.xrl  absinthe_parser.yrl
```
- Compile and clean the project: `mix compile && mix clean --deps`
- The `.erl` files should be gone after cleaning, but they are not:
```bash
$ ls deps/absinthe/src/
absinthe_lexer.erl  absinthe_lexer.xrl  absinthe_parser.erl absinthe_parser.yrl
```

### Environment

* Elixir & Erlang/OTP versions (elixir --version): 
```bash
$ elixir --version
Erlang/OTP 21 [erts-10.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe] [dtrace]

Elixir 1.7.3 (compiled with Erlang/OTP 21)
```
* Mix version:
```bash
$ mix --version
Erlang/OTP 21 [erts-10.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe] [dtrace]

Mix 1.7.3 (compiled with Erlang/OTP 21)
```
* Operating system: 
macOS Mojave 10.14

**Comment by ggcampinho** [comment]
I was checking the `clean` task and it uses a simple approach of cleaning the `_build/deps` folder instead of invoking the `deps.clean` task [like it does for the project](https://github.com/elixir-lang/elixir/blob/master/lib/mix/lib/mix/tasks/clean.ex#L27-L30). Would make sense to call this task?

If so, I also notice that `deps.clean` [does a similar thing](https://github.com/elixir-lang/elixir/blob/master/lib/mix/lib/mix/tasks/deps.clean.ex#L106-L110). Would make sense to change `deps.clean` first to consider those files?

**Comment by f355** [comment]
@ggcampinho unfortunately I'm not familiar with the codebase enough to definitely answer your questions (assuming they are for me :)), but from a common sense point of view it would be much better to use the same logic for the dependencies as for the project itself.

Also, the dependencies might be built by something other than Mix, e.g. rebar, and probably should be cleaned using that as well to accommodate for custom scripts/hooks, so maybe the cleaning logic should mirror the dependency compilation.

**Comment by josevalim** [comment]
@ggcampinho yes, I think we need to traverse the deps but we need to be careful with some things:

1. We can only invoke the mix task for mix dependencies
2. We need to make sure before invoking the clean task, the dependency is OK, otherwise cleaning can fail or we may even end-up trying to compile the dependency only to clean it

As long as we have those in mind, we should be good. I would start with changing deps.clean and then we can figure out a way of changing `clean` too.

**Comment by michalmuskala** [comment]
I think a simple approach might be to change how the `leex` and `yecc` compilers work in Elixir - instead of writing the files to `.src` and then letting the `erlang` compiler do the job, we could have them write the temporary `.erl` file somewhere in `_build` and compile to `.beam` directly. This means we wouldn't have any intermediary files in source directories.

**Comment by josevalim** [comment]
I am on this!

**Comment by josevalim** [comment]
Ok, I am not on this. I think the solution of changing the compiler is not going to fly.

We probably need to change deps.clean instead, as @ggcampinho first suggested. @ggcampinho are you planning to tackle this?

**Comment by drincruz** [comment]
Just reading through this thread, I gave it a go on removing those `.erl` files that get built in `deps/{package}/src`. [This](https://github.com/elixir-lang/elixir/compare/master...drincruz:bugfix/mix/clean_deps?expand=1) works, but also does *not* sound like the solution @josevalim outlined [here](https://github.com/elixir-lang/elixir/issues/8262#issuecomment-428479538). 

>We need to make sure before invoking the clean task, the dependency is OK, otherwise cleaning can fail or we may even end-up trying to compile the dependency only to clean it

how do we accomplish making sure the dependency is OK? 

Just thought I would try and give this a go!

**Comment by josevalim** [comment]
@drincruz right. We cannot hardcode the "src" directory because any dependency may have multiple compilers and any other directory can have compiled artefacts. It may be that the best way to solve this is to actually simplify the depending cleaning into two: "build" and "source" and it is always all or nothing, instead of being granular.

**Comment by josevalim** [comment]
Another suggestion given by @michalmuskala in #10153 is to emit the .beam files directly from .yrl/.xrl compilers.

**Comment by josevalim** [comment]
@michalmuskala I looked into your solution and it has complexities too. Namely, the .yrl and .xrl files may depend on other files as behaviours and parse transforms, and we would need to compile those first. It is hard for this to happen in practice but it is technically a limitation (and a backwards incompatible change).

`deps.clean` may be the best option here after all... it will also be the most correct one for anyone implementing other compilers.

**Comment by Ch4s3** [comment]
@josevalim I looked at a few possible designs here and it is deceptively hard because the generated files could be in non standard places. I'm happy to take a shot at a different approach, I'm just shrugging to find all of the rough edges oft he problem.

**Comment by nseaSeb** [comment]
Having followed this thread, I'd like to offer one observation and a concrete option that I don't think has been fully explored.

Observation: for the current project, mix clean already solves this generically. Each compiler's clean/0 is manifest-driven: Mix.Compilers.Erlang.clean/1 deletes every file recorded at compile time, wherever it lives — including a .erl emitted into src/. No hardcoded paths or extensions. The gap is only that dependencies never get this treatment: mix clean --deps and mix deps.clean remove directories wholesale and never consult the dep's compiler manifests, so an in-source generated .erl survives exactly when sources are kept.

Option: cold-read the Mix-owned manifests. All built-in compilers that can emit into the source tree (compile.erlang, compile.yecc, compile.leex) share a single manifest format owned by Mix itself, stored under _build/*/lib/<dep>/.mix/. Before wiping the build directories, dep cleaning could read those manifests as data — no dep project loading, no task invocation, no risk of compiling a dep just to clean it (the caveats raised earlier) — take the union across build envs, keep only entries contained within that dep's source directory, and delete those files.
 
Properties worth noting:

 - No knowledge of .xrl/.yrl or any path layout is encoded anywhere — we only delete what was recorded as generated at compile time, which I believe addresses the brittleness objection to #15318.
 - Strictly better, never worse: rebar deps and third-party compilers behave exactly as today.
 - Ordering matters: the sweep must run before _build removal, which also makes repeated cleans a no-op.
 - The containment check (entry must be inside the dep's directory) bounds the blast radius, and manifests are locally produced, so no new trust surface.

Known limits: third-party source dirs are not covered(that would need loading the dep's project, or a future opt-in compiler callback along the lines of "list ymanifests already destroyed by hand (rm -rf _build) leave orphans — same as today.

I don't know if this helps? Perhaps it sums up the problem?

**Comment by josevalim** [comment]
@nseaSeb according to our contributions guidelines, please don’t post AI generated content in discussions. You can write in your native language and ask it to translate if there is a language barrier.

**Comment by nseaSeb** [comment]
Ok sorry for this @josevalim
